### PR TITLE
Support submission-time image override

### PR DESCRIPTION
**Pull request by fuhuifang**
- [x] Image override
- [x] dict override (e.g. tags)
- [x] Allow yaml key in `tenant_overrides`, e.g. "personal" instead of the true tenant id
- [x] Unit tests
- [x] [Successful exp](https://ml.azure.com/runs/3d213322-9dc7-409d-aaca-2cdbfc909fee?wsid=/subscriptions/48bbc269-ce89-4f6f-9a12-c6f91fcb772d/resourcegroups/aml1p-rg/workspaces/aml1p-ml-wus2&tid=72f988bf-86f1-41af-91ab-2d7cd011db47): The original yaml files are renamed to "xxx.not_used" to differentiate.
image

### CODE REVIEW

File: requirements/requirements-pipeline.txt [hunk 1]
```diff
@@ -6,7 +6,8 @@ azure-cli-core==2.22.1
 ruamel.yaml~=0.16
 gitpython~=3.1
 hydra-core==1.0.3
-flatten-dict==0.4.0 
+flatten-dict==0.4.0
```
Comment (dkmiller) [inline requirements/requirements-pipeline.txt:9]: Nit: these requirements should be of the form `flatten-dict~=0.4`.

Only put "pinned" dependency versions in the dev requirements.txt

**Review by haozhestat — COMMENTED** [review]
This draft PR looks good to me. I don't have specific comments now. Feel free to proceed with unit tests. Thanks for adding this fascinating feature.

I will come back to review again after this draft is published.

**Review by haozhestat — APPROVED** [review]
After this PR is merged, could you add some documentations on the image override.

**Review by dkmiller — APPROVED** [review]
Great, comprehensive, unit tests.

**Pull request merged**